src/debugging/python/PythonDebugHelper.ts

```diff
@@ -0,0 +1,196 @@
+                    {
+                        /* eslint-disable-next-line no-template-curly-in-string */
+                        localRoot: "${workspaceFolder}",
+                        remoteRoot: "/app"
+                    }
+                ],
+                projectType: options.projectType
+            }
+        }];
+
+        // If we generated compose files, then we should generate a Python attach configuration.
+        if (context.generateComposeTask) {
+            configs.push(
+                {
+                    name: "Python: Remote Attach",
+                    type: "python",
+                    request: "attach",
+                    host: "localhost",
+                    port: PythonDefaultDebugPort,
+                    pathMappings: [
+                        {
+                            /* eslint-disable-next-line no-template-curly-in-string */
+                            localRoot: "${workspaceFolder}",
+                            remoteRoot: "/app"
+                        }
+                    ]
+                });
+        }
+
+        return configs;
+    }
+
+    public async resolveDebugConfiguration(context: DockerDebugContext, debugConfiguration: PythonDockerDebugConfiguration): Promise<ResolvedDebugConfiguration | undefined> {
+        const containerName = inferContainerName(debugConfiguration, context, context.folder.name);
+
+        // Since Python is a special case, we need to ensure the container is removed before attempting to resolve
+        // the debug configuration.
+        try {
+            await this.cliDockerClient.removeContainer(containerName, { force: true });
+        } catch {}
+
+        const debuggerLogFilePath = PythonExtensionHelper.getDebuggerLogFilePath(context.folder.name);
+        await fse.remove(debuggerLogFilePath);
+
+        let debuggerReadyPromise = new Promise((resolve) => resolve());
+        if (debugConfiguration.preLaunchTask) {
+            // There is this limitation with the Python debugger where we need to ensure it's ready before allowing VSCode to attach,
+            // if attach happens too soon then it will fail silently. The workaround here is to set the preLaunchTask to undefined,
+            // then execute it ourselves with a listener to when it is finished, then wait for the debugger to be ready and return
+            // the resolved launch configuration.
+
+            const task = await this.tryGetPrelaunchTask(debugConfiguration.preLaunchTask);
+
+            if (!task) {
+                throw new Error(`Unable to find the prelaunch task with the name: ${debugConfiguration.preLaunchTask}`);
+            }
+
+            debugConfiguration.preLaunchTask = undefined;
+
+            /* eslint-disable-next-line @typescript-eslint/no-floating-promises */
+            vscode.tasks.executeTask(task);
+
+            debuggerReadyPromise = PythonExtensionHelper.ensureDebuggerReady(task, debuggerLogFilePath, containerName, this.cliDockerClient);
```

Comment:
Since this subscribes to the task end event, should it be created *before* the task starts executing?

Reply:
Yep, make sense.